lightstep/b3_propagator.py

```diff
@@ -0,0 +1,149 @@
+            warn(
+                "If not propagating only the sampling state, traceid and "
+                "spanid must be defined"
+            )
+
+        carrier.update(baggage)
+
+        if traceid is not None:
+            # FIXME This adds zeroes on the MS side until traceid is 128b long.
+            # Confirm that this is the desired implementation.
+            carrier[_TRACEID] = format(traceid, "032x")
+        if spanid is not None:
+            carrier[_SPANID] = format(spanid, "016x")
+
+    def extract(self, carrier):
+
+        case_insensitive_carrier = {}
+        for key, value in carrier.items():
+            for b3_key in [
+                _SINGLE_HEADER,
+                _TRACEID,
+                _SPANID,
+                _PARENTSPANID,
+                _SAMPLED,
+                _FLAGS,
+            ]:
+                if re_compile(key, IGNORECASE).match(b3_key):
+                    case_insensitive_carrier[b3_key] = value
+                else:
+                    case_insensitive_carrier[key] = value
+
+        carrier = case_insensitive_carrier
+        baggage = {}
+
+        if _SINGLE_HEADER in carrier.keys():
+            fields = carrier.pop(_SINGLE_HEADER).split("-", 4)
+            baggage.update(carrier)
+            len_fields = len(fields)
+            if len_fields == 1:
+                sampled = fields[0]
+            elif len_fields == 2:
+                traceid, spanid = fields
+            elif len_fields == 3:
+                traceid, spanid, sampled = fields
+            else:
+                traceid, spanid, sampled, parent_spanid = fields
+                baggage[_PARENTSPANID] = int(parent_spanid, 16)
+            if sampled == "d":
+                baggage[_FLAGS] = 1
+            else:
+                baggage[_SAMPLED] = int(sampled, 16)
+        else:
+            traceid = carrier.pop(_TRACEID, None)
+            spanid = carrier.pop(_SPANID, None)
+            parentspanid = carrier.pop(_PARENTSPANID, None)
+            sampled = carrier.pop(_SAMPLED, None)
+            flags = carrier.pop(_FLAGS, None)
+
+            if sampled is flags is (traceid and spanid) is None:
+
+                raise SpanContextCorruptedException()
+
+            if parentspanid is not None:
+                baggage[_PARENTSPANID] = int(parentspanid, 16)
+
+            if flags == 1:
+                baggage[_FLAGS] = flags
+                if sampled is not None:
+                    warn(
+                        "x-b3-flags: 1 implies x-b3-sampled: 1, ignoring "
+                        "the received value of x-b3-sampled"
+                    )
+            elif sampled is not None:
+                baggage[_SAMPLED] = int(sampled, 16)
+
+            baggage.update(carrier)
+
+            if baggage == OTSpanContext.EMPTY_BAGGAGE:
+                baggage = None
+
+        # FIXME This truncates the MSb of traceid until only 64b are left.
```

Comment:
the implementation here should leave the trace ID as is when it's propagated to ensure it matches what was passed in. the truncation should only occur when reporting to lightstep

Reply:
Sure, I'll change it. What do you refer to with `when reporting to lightstep`?

Reply:
A good example of this is in the ruby implementation, where a separate field stores the 128-bit vs the 64-bit trace IDs, to remain compatible with the current propagators: https://github.com/lightstep/lightstep-tracer-ruby/pull/86/files#diff-d87db7512b8e96c0a8d9463d52de94beR13-R14. This ensures existing propagation continues to work as is.